rtengine/ipwavelet.cc

```diff
@@ -2205,7 +2203,6 @@ void ImProcFunctions::ip_wavelet(LabImage * lab, LabImage * dst, int kall, const
 
             Evaluate2(*LdecompLL, meang, meanNg, sigmag, sigmaNg, MaxPg, MaxNg, wavNestedLevels);
             Evaluate2(*Ldecompdst, mean, meanN, sigma, sigmaN, MaxP, MaxN, wavNestedLevels);
-            float sig = 2.f;
```

Comment:
While it does not change the resulting values, I think readability is reduced by inlining this instead of e.g. making it a `constexpr`. I am now unable to see that this is always the "same two" in the following places where it was inlined.

I understand this would prevent replacing `pow` by `sqrt(sqrt(..))` but I think it is not obvious that the gain speed is wort the loss of readability. (IIRC, compilers even replace `pow(x, y)` by optimized variants for specific `y` even though I am not sure if `y=2^{-n}` would qualify.)

Reply:
It seems like this computation is not done in a loop, e.g. per-pixel, so speed gains are probably inconsequential.

Reply:
> It seems like this computation is not done in a loop, e.g. per-pixel, so speed gains are probably inconsequential.

The pow() was inside a pixel loop

```
                        for (int i = 0; i < Wlvl_L * Hlvl_L; i++) {
```

Reply:
> The pow() was inside a pixel loop

Yes, you are right. So it is not inconsequential. Just a trade-off.

Reply:
> IIRC, compilers even replace pow(x, y) by optimized variants for specific y

I just checked using godbolt. At least gcc does a call to std::pow in this case

Reply:
while a sqrt(sqrt( is just two branch free instructions

Reply:
> while a sqrt(sqrt( is just two branch free instructions

I am not arguing that the current code will reliably be as a fast as the proposed change. I am questioning whether loosing the information that `sig` is a parameter of the algorithm that is used in multiple places is worth the gain. (That both of us do not seem to have measured ATM.)

Reply:
![grafik](https://user-images.githubusercontent.com/13733487/109058156-f03abd80-76e2-11eb-96b9-3f871cddd96a.png)

Reply:
> (That both of us do not seem to have measured ATM.)

I will measure it tomorrow. I guess it will speedup a little bit as the new code is auto-vectorized by the compiler while the old code (calling std::pow()) is not

Reply:
I am not sure which point the screenshot is making? You mean that there is a `call` in there? I already agreed that this can happen even if `sig` is a `constexpr`. We are still guessing as to how much faster this ends up being on actual images.

It is also not the point I am personally trying to make. I think we should acknowledge that information is lost with this code change and that change this algorithm in the future will get harder. I would be fine with saying that the two instructions versus the call is definitely worth it, but a screenshot does not do that.

Reply:
> I am not sure which point the screenshot is making?

The point is that a call to std::pow will not be vectorized by -O3

But the std::sqrt can be vecorized (as you can see in the `sqrtps` instructions in assembly, which processes 4 floats at once...

Reply:
we should simply ask @Desmis about the meaning of sig = 2.f in this code...

Reply:
> But the std::sqrt can be vecorized (as you can see in the sqrtps instructions in assembly, which processes 4 floats at once...

So is this the point where you would say that the loss of information is definitely worth it?

Reply:
@adamreichold @heckflosse You/we could start a more general discussion on coding standards and when to split hairs. Could be worthwhile. There is even a tool for that in GitHub https://github.com/Beep6581/RawTherapee/discussions

Reply:
> So is this the point where you would say that the loss of information is definitely worth it?

Without knowing about the informational value of `sig` we don't know ig there is a loss of information at all.
Let's wait for @Desmis reply and not split hairs...

Reply:
Personally, I would be most happy with continuing along the way of

> Also, my first cppcheck fix for this file didn't contain this optimization at all. I just wanted to make the file cppcheck clean...

Making the file cppcheck-clean does not require dropping `sig`, inquiring the original author, performing benchmarks or changing coding standards. I think we could merge this quickly.

Reply:
> Without knowing about the informational value of sig we don't know if there is a loss of information at all.
Let's wait for @Desmis reply and not split hairs...

Waiting is fine with me. Concerning hair, I actually did not see that second comment before replying to its predecessor.

Reply:
> Making the file cppcheck-clean does not require dropping sig, inquiring the original author, performing benchmarks or changing coding standards. I think we could merge this quickly.

I agree. I didn't even get rid of sig in my first post...